### ISSUE TITLE
Tiles not created only ParseError (Exception: Unknown tile tag: 15)

### ISSUE THREAD

**Issue by Naphier** [mapbox/mapbox-unity-sdk#45]
## System Info
Unity v5.5.2 (64-bit Windows editor)
Mapbox Unity SDK  v0.4 beta

## Background
I'm able to run example scenes in a project with just mapbox in it so I'm thinking there's a collision with something somewhere, but can't find it. I had `ICSharpCode.SharpZipLib` in the project and removed my existing one in favor of the one that cam with mapbox. I'm also using NewtonSoft's json, I tried removing that and no difference.

## Issue
When running MapController no valid vector tile data is returned from the request. All I get are ParseErrors (with no further helpful information). After some digging I was able to glean the following:
- VectorTile.ParseData throws an exception "Unknown tile tag: 15"
- Response byte array is of length 44
- Bytes returned by the response are the same bytes that are returned from Compression.Decompress.
- MapController inspector settings are ![capture](https://cloud.githubusercontent.com/assets/10132540/25634746/be6147fa-2f49-11e7-973f-4e502271d646.png)
- Editor log: [Editor.txt](https://github.com/mapbox/mapbox-unity-sdk/files/971736/Editor.txt)

## Plugins in the project:
Native file browser:
- System.Windows.Forms
- Ookii.Dialogs
- StandaloneFileBrowser.jslib

TextMesh pro:
- TextMeshPro-Editor
- TextMeshPro
- TMPro_Plugin
- TextMeshPro
- vcomp120

Mapbox:
- Mapbox.VectorTile.ExtensionMethods
- Mapbox.VectorTile.Geometry
- Mapbox.VectorTile.PbfReader
- ICSharpCod.SharpZipLib
- Mapbox.Json
- Triangulation

Other:
- Newtonsoft.Json

**Comment by david-rhodes** [comment]
@Naphier Thanks for the report! I have an idea what the issue is, but you can help me confirm. In your `HTTPRequest` class, please log the `url`. Copy the url and paste it into your web browser. Does the data look complete? 

If the data looks correct in your web browser, you will need to check that all of your factories are using a proper `MapID`. It seems in some circumstances, the MapID is does not maintain serialization across projects or is corrupted by Unity version differences.

**Comment by ahokinson** [comment]
I'm using the MapID from MapboxStudio:

<img width="1067" alt="screen shot 2017-05-02 at 4 12 40 pm" src="https://cloud.githubusercontent.com/assets/1762048/25637213/3cde6664-2f52-11e7-9a2a-6d43aedbc6e5.png">

![screen shot 2017-05-02 at 4 13 36 pm](https://cloud.githubusercontent.com/assets/1762048/25637278/6ec085ea-2f52-11e7-914f-93af67e210a9.png)
![screen shot 2017-05-02 at 4 14 16 pm](https://cloud.githubusercontent.com/assets/1762048/25637282/70418e96-2f52-11e7-8703-178ca86891dd.png)

And I'm getting the same issue under the same setup as @Naphier.

![screen shot 2017-05-02 at 4 09 58 pm](https://cloud.githubusercontent.com/assets/1762048/25637309/81e7a4fa-2f52-11e7-97f5-e7e66d79ff90.png)

When debugging the URL, I get:
`https://api.mapbox.com:443/v4/ahokinson.8ww3j897/9/157/190.vector.pbf`
Where I have excluded the access token.

**Comment by ahokinson** [comment]
I should note that my map tiles are created, but in trying to use custom POI data, I get these errors.

**Comment by david-rhodes** [comment]
@Naphier @ahokinson When you paste the url into your browser, does it download a .pbf file? When testing `https://api.mapbox.com:443/v4/ahokinson.8ww3j897/9/157/190.vector.pbf` I get:

```
{
  "message": "Tile does not exist"
}
```

When using my access token.

**Comment by Naphier** [comment]
After adding in a Debug.Log message into the HTTPRequest's constructor everything I am able to get the maps. When I use the DriveVisualization (map visualisation) everything seems to be OK.  Maybe it's something to do with async and Unity and the Debug.Log message is ensuring the main thread completes a cycle before starting the DoRequest coroutine?

.

My URLS show an image in the browser. 

.

So if I have a range of 1,1,1,1 it now works OK. But if I do a range of 10,10,10,10 it will get me the tiles then start throwing index out of range exceptions.
```

IndexOutOfRangeException: Array index is out of range.
Mapbox.VectorTile.VectorTileReader..ctor (System.Byte[] data, Boolean validate)
Mapbox.VectorTile.VectorTile..ctor (System.Byte[] data, Boolean validate)
Mapbox.Map.VectorTile.ParseTileData (System.Byte[] data) (at Assets/Mapbox/Core/Map/VectorTile.cs:106)
UnityEngine.Debug:LogException(Exception)
Mapbox.Map.VectorTile:ParseTileData(Byte[]) (at Assets/Mapbox/Core/Map/VectorTile.cs:112)
Mapbox.Map.Tile:HandleTileResponse(Response) (at Assets/Mapbox/Core/Map/Tile.cs:155)
Mapbox.Unity.<DoRequest>c__Iterator0:MoveNext() (at Assets/Mapbox/Core/Unity/HTTPRequest.cs:40)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
```

```
vectorTile.Error: ParseError    responseData: 20
parameters: Id: 16/18831/24226  MapId: mapbox.mapbox-traffic-v1,mapbox.mapbox-streets-v7  Fs: Mapbox.Unity.FileSource
UnityEngine.Debug:LogFormat(String, Object[])
Mapbox.MeshGeneration.Factories.<CreateMeshes>c__AnonStorey1:<>m__0() (at Assets/Mapbox/Scripts/MeshGeneration/Factories/MeshFactory.cs:89)
Mapbox.Map.Tile:HandleTileResponse(Response) (at Assets/Mapbox/Core/Map/Tile.cs:162)
Mapbox.Unity.<DoRequest>c__Iterator0:MoveNext() (at Assets/Mapbox/Core/Unity/HTTPRequest.cs:40)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
```

**Comment by ahokinson** [comment]
@david-rhodes I do see that now...  

Given these factories, do you have any suggestions?

![screen shot 2017-05-02 at 4 42 47 pm](https://cloud.githubusercontent.com/assets/1762048/25638444/6fad5e0c-2f56-11e7-98a5-b91f17446ef6.png)
![screen shot 2017-05-02 at 4 43 22 pm](https://cloud.githubusercontent.com/assets/1762048/25638458/789551e6-2f56-11e7-909d-cbd4f3a2dba1.png)
![screen shot 2017-05-02 at 4 43 37 pm](https://cloud.githubusercontent.com/assets/1762048/25638469/8527e61c-2f56-11e7-86c4-2453bff3e272.png)
![screen shot 2017-05-02 at 4 43 59 pm](https://cloud.githubusercontent.com/assets/1762048/25638481/90e14692-2f56-11e7-82ae-1caff4f74b07.png)
![screen shot 2017-05-02 at 4 44 18 pm](https://cloud.githubusercontent.com/assets/1762048/25638492/9a038fbe-2f56-11e7-9267-f73da0b6d1ed.png)

What I find odd is that using the example dataset works.  I uploaded a geojson file to MapBox studio to test with.  It looks okay to me, but maybe I've used Studio completely wrong.  That's a strong possibility...

<img width="1183" alt="screen shot 2017-05-02 at 4 45 42 pm" src="https://cloud.githubusercontent.com/assets/1762048/25638568/dcacaa8a-2f56-11e7-9c2b-f092b71d06ea.png">

**Comment by david-rhodes** [comment]
@Naphier 

>My URLS show an image in the browser.

You should not see an image for a vector tile request. Can you post a url without your access token? It doesn't look like you are using a custom tileset, but my best guess is that data is missing for a tile somewhere in that larger range.

**Comment by david-rhodes** [comment]
@ahokinson My guess is that your data does not exist for the tiles you are requesting. Have you tried a different region or zoom level?

**Comment by ahokinson** [comment]
@david-rhodes Trying just now yields the same results with different amounts of ParseErrors.  Even if there was no data for a tile, would a ParseError really be expected?

![screen shot 2017-05-02 at 4 55 50 pm](https://cloud.githubusercontent.com/assets/1762048/25638978/3a1960b8-2f58-11e7-9b1a-acf640ad8144.png)

Here is everything working with the standard poi_label set from mapbox.mapbox-streets-v7
![screen shot 2017-05-02 at 5 00 55 pm](https://cloud.githubusercontent.com/assets/1762048/25639206/f914785e-2f58-11e7-842e-893b3fb8503a.png)

Is there something required of datasets to work with the PoiMesh system?

**Comment by david-rhodes** [comment]
@ahokinson 

> Even if there was no data for a tile, would a ParseError really be expected

Well, I'm not sure that we are handling a case of "tile missing" at this moment. Probably just an oversight on our part. So, if the data (which is not a valid .pbf at this point) cannot be parsed as expected, that's the error you will see.

> Is there something required of datasets to work with the PoiMesh system?

No. The root of this problem is that your url requests are not valid. Are you able to share your tileset with me? If you add that tileset to a studio style, does it render as you would expect?

**Comment by ahokinson** [comment]
Yeah, I definitely agree with you on the first point.  I just thought you were implying this might be normal.  My mistake.

The data I'm using for experimenting is as follows:
[cities.txt](https://github.com/mapbox/mapbox-unity-sdk/files/971985/cities.txt)

I'm willing to share the tileset, but I'm not sure how.  When I attach it to a style, it appears correctly:
<img width="1399" alt="screen shot 2017-05-02 at 5 12 47 pm" src="https://cloud.githubusercontent.com/assets/1762048/25639650/a4b009de-2f5a-11e7-9c74-2ec63df5d397.png">

**Comment by david-rhodes** [comment]
@ahokinson Ok, I think I've solved the problem.

I uploaded your data to my own tileset and added it's ID to the mesh factory. When requesting your tile coordinates, I also got "no tile exists." But, when requesting the San Francisco area, for example, I do see POIs being created and the url request is valid (downloads a pbf and everything). 

That said, I did have to remove:

```cs
            if (!feature.Properties.ContainsKey("name"))
                return;
```
and
```cs
            var tm = go.GetComponent<ILabelVisualizationHelper>();
            tm.Initialize(feature.Properties);
```

from `PoiVisualizer.cs` to be compatible with this dataset. The reason is because of an early return since `name` was not a valid feature in your set.

The `ParseError` is a generic error, which I just realized is a bug. We actually have more robust error reporting available. But, this occurs when requesting a tile that does not contain any data. I'm assuming our servers do not generate tiles that are not needed (as data does not exist). Therefore, we cannot parse a "null" tile. :smirk:

Anyway, change your code as above and you should see some POIs.

**Comment by ahokinson** [comment]
I made the first change, but not the second because I implemented the ILabelVisualizationHelper interface which is actually a very convenient spot to initialize my objects. :)

And sure enough, I can see markers!

![screen shot 2017-05-02 at 6 27 51 pm](https://cloud.githubusercontent.com/assets/1762048/25642042/530ca21c-2f65-11e7-8073-9b0948225459.png)

Thanks for persisting with me on this one!  Everything seems to make sense now, but it wouldn't have without your responses.  Glad to hear robust error handling will be coming in the future.

**Comment by david-rhodes** [comment]
Awesome! That makes sense--most people don't implement `ILabelVisualizationHelper`, so that's why I suggested its removal.

We need developers like you to help find bugs and suggest improvements. However, please have patience as we are still in beta.

**Comment by ahokinson** [comment]
Lol I'm sorry.  I don't mean to sound impatient.  It's the frustration of being mostly correct, but missing something very small.  I absolutely love this sdk and the potential it offers.  It could not have been made public at a better time.  Thanks for your hard word.  Seriously, it's much appreciated!

**Comment by david-rhodes** [comment]
Thanks @ahokinson! Please continue to provide feedback or even contribute with your own improvements! 😄

**Comment by david-rhodes** [comment]
@Naphier Did any of the above help you? If not, which tiles specifically are not being created?

**Comment by Naphier** [comment]
After a bit of goofing around I think that my issue was actually an invalid API token. I removed a character from the token to test this out and I'm getting the same parse error as before. So perhaps it wasn't being serialized properly or maybe I made a mistake?! 

It does seem to be fixed now and everything's working. Try putting in a bad API key and actually logging the exception from VectorTile.ParseTileData in the existing try-catch and you'll see the "Unkown Tile tag : 15" exception. I think the easiest solution for this is to just verify the token before initiating the tile data requests. But then again, maybe the new method of using the token that you mentioned in #46 will make this less of an issue. 

Thanks for the quick support!
